staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -746,7 +747,33 @@ func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, pred
 		}
 		return nil, readResourceVersion, "", nil
 	}
-	return c.watchCache.WaitUntilFreshAndList(ctx, listRV, pred.MatcherIndex())
+	return c.watchCache.WaitUntilFreshAndList(ctx, listRV, matcherIndex(ctx, pred))
+}
+
+func matcherIndex(ctx context.Context, pred storage.SelectionPredicate) []storage.MatchValue {
+	re, _ := request.RequestInfoFrom(ctx)
+	if re == nil || !re.IsResourceRequest || re.Subresource != "" || !whitelistResource(re.Resource) {
```

Comment:
We shouldn't hardcode resources here - instead you should do that purely relying on whethever `IndexFields` contains `metadata.namespace` inside.